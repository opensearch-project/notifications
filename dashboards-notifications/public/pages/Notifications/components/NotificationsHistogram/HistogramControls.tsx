/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { HISTOGRAM_TYPE } from '../../../../utils/constants';
import { EuiSuperSelect, EuiSuperSelectOption } from '@elastic/eui';
import React from 'react';

interface HistogramControlsProps {
  histogramType: keyof typeof HISTOGRAM_TYPE;
  setHistogramType: (histogramType: keyof typeof HISTOGRAM_TYPE) => void;
}

export function HistogramControls(props: HistogramControlsProps) {
  const histogramTypeOptions: Array<EuiSuperSelectOption<
    keyof typeof HISTOGRAM_TYPE
  >> = Object.entries(HISTOGRAM_TYPE).map(([value, inputDisplay]) => ({
    value: value as keyof typeof HISTOGRAM_TYPE,
    inputDisplay,
  }));

  return (
    <>
      <EuiSuperSelect
        style={{ width: 300 }}
        prepend="Slice by"
        fullWidth
        options={histogramTypeOptions}
        valueOfSelected={props.histogramType}
        onChange={props.setHistogramType}
      />
    </>
  );
}
