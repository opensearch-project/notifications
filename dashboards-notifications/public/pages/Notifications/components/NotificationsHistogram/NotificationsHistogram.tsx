/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  Axis,
  BarSeries,
  Chart,
  Datum,
  niceTimeFormatByDay,
  ScaleType,
  Settings,
  timeFormatter,
} from '@elastic/charts';
import { euiPaletteColorBlind, EuiSpacer } from '@elastic/eui';
import _ from 'lodash';
import React from 'react';
import {
  ContentPanel,
  ContentPanelActions,
} from '../../../../components/ContentPanel';
import { HISTOGRAM_TYPE } from '../../../../utils/constants';
import { HistogramControls } from './HistogramControls';

interface NotificationsHistogramProps {
  histogramType: keyof typeof HISTOGRAM_TYPE;
  setHistogramType: (histogramType: keyof typeof HISTOGRAM_TYPE) => void;
  histogramData: Array<Datum>;
}

export function NotificationsHistogram(props: NotificationsHistogramProps) {
  const formatter = timeFormatter(niceTimeFormatByDay(1));
  return (
    <>
      <ContentPanel
        actions={
          <ContentPanelActions
            actions={[
              {
                component: (
                  <HistogramControls
                    histogramType={props.histogramType}
                    setHistogramType={props.setHistogramType}
                  />
                ),
              },
            ]}
          />
        }
        bodyStyles={{ padding: 'initial' }}
        title={`Notifications by ${_.get(
          HISTOGRAM_TYPE,
          props.histogramType,
          Object.values(HISTOGRAM_TYPE)[0]
        ).toLowerCase()}`}
        titleSize="m"
      >
        <EuiSpacer />
        <Chart size={{ height: 250 }}>
          <Settings
            theme={{
              colors: {
                vizColors: euiPaletteColorBlind({ sortBy: 'natural' }),
              },
            }}
            showLegend={true}
          />
          <BarSeries
            id="status"
            name="Status"
            data={props.histogramData}
            xAccessor={'x'}
            yAccessors={['y']}
            splitSeriesAccessors={['g']}
            stackAccessors={['g']}
            xScaleType={ScaleType.Time}
          />
          <Axis id="bottom-axis" position="bottom" tickFormat={formatter} />
          <Axis id="left-axis" position="left" showGridLines />
        </Chart>
      </ContentPanel>
    </>
  );
}
