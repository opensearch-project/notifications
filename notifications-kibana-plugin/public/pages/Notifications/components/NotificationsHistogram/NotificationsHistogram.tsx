/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  Axis,
  BarSeries,
  Chart,
  DataGenerator,
  Datum,
  Settings,
} from '@elastic/charts';
import { euiPaletteColorBlind, EuiSpacer } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import {
  ContentPanel,
  ContentPanelActions,
} from '../../../../components/ContentPanel';
import { HISTOGRAM_TYPE } from '../../../../utils/constants';
import { HistogramControls } from './HistogramControls';

interface NotificationsHistogramProps {}

export function NotificationsHistogram(props: NotificationsHistogramProps) {
  const [histogramType, setHistogramType] = useState<
    keyof typeof HISTOGRAM_TYPE
  >('CHANNEL_TYPE');
  const [data, setData] = useState<Datum[]>([]);

  useEffect(() => {
    const dg = new DataGenerator();
    const data = dg.generateGroupedSeries(25, 2, 'Channel-')
    data[18].y = 18;
    setData(data);
  }, []);

  return (
    <>
      <ContentPanel
        actions={
          <ContentPanelActions
            actions={[
              {
                component: (
                  <HistogramControls
                    histogramType={histogramType}
                    setHistogramType={setHistogramType}
                  />
                ),
              },
            ]}
          />
        }
        bodyStyles={{ padding: 'initial' }}
        title={`Notifications by ${HISTOGRAM_TYPE[
          histogramType
        ].toLowerCase()}`}
        titleSize="m"
      >
        <EuiSpacer />
        <Chart size={{ height: 250 }}>
          <Settings
            theme={{
              colors: {
                vizColors: euiPaletteColorBlind({ sortBy: 'natural' }),
              },
            }}
            showLegend={true}
          />
          <BarSeries
            id="status"
            name="Status"
            data={data}
            xAccessor={'x'}
            yAccessors={['y']}
            splitSeriesAccessors={['g']}
            stackAccessors={['g']}
          />
          <Axis id="bottom-axis" position="bottom" />
          <Axis id="left-axis" position="left" showGridLines />
        </Chart>
      </ContentPanel>
    </>
  );
}
