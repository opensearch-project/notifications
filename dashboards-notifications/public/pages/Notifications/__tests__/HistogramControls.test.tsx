/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from '@testing-library/react';
import React from 'react';
import { HistogramControls } from '../components/NotificationsHistogram/HistogramControls';

describe('<HistogramControls /> spec', () => {
  it('renders the component', () => {
    const setHistogramType = jest.fn();
    const utils = render(
      <HistogramControls
        histogramType="CHANNEL_TYPE"
        setHistogramType={setHistogramType}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
