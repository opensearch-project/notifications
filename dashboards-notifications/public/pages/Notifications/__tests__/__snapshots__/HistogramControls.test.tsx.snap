// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistogramControls /> spec renders the component 1`] = `
<div
  class="euiPopover euiPopover--anchorDownCenter euiPopover--displayBlock euiSuperSelect euiSuperSelect--fullWidth"
>
  <div
    class="euiPopover__anchor"
  >
    <input
      type="hidden"
      value="CHANNEL_TYPE"
    />
    <div
      class="euiFormControlLayout euiFormControlLayout--fullWidth euiFormControlLayout--group"
    >
      <label
        class="euiFormLabel euiFormControlLayout__prepend"
      >
        Slice by
      </label>
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <span
          class="euiScreenReaderOnly"
          id="random_html_id"
        >
          Select an option: Channel type, is selected
        </span>
        <button
          aria-haspopup="true"
          aria-labelledby="undefined random_html_id"
          class="euiSuperSelectControl euiSuperSelectControl--fullWidth euiSuperSelectControl--inGroup"
          style="width: 300px;"
          type="button"
        >
          Channel type
        </button>
        <div
          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
        >
          <span
            class="euiFormControlLayoutCustomIcon"
          >
            <svg
              aria-hidden="true"
              class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
              focusable="false"
              height="16"
              role="img"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
